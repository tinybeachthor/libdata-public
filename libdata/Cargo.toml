[package]
name = "libdata"
version = "0.1.0"
authors = ["Martin Toman <toman.martin@live.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
datacore = { path = "../datacore" }
protocol = { path = "../protocol" }
random-access-storage = { path = "../random-access-storage" }
anyhow = "1.0.44"
rand = { version = "0.7.3", features = [ "std", "wasm-bindgen" ] }
rand_chacha = "0.3.1"
blake3 = "1.3.1"
futures-lite = "1.12.0"
async-std = "1.10.0"
async-trait = "0.1.24"
async-channel = "1.6.1"
hex = "0.4"

[dev-dependencies]
random-access-memory = { path = "../random-access-memory" }
quickcheck = "0.9.2"
insta = "1.8.0"
async-std = { version = "1.10.0", features = ["attributes"] }
sluice = "0.5.5"
