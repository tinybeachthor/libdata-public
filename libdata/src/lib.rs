#![forbid(unsafe_code, bad_style, nonstandard_style, future_incompatible)]
#![forbid(rust_2018_idioms, rust_2021_compatibility)]
#![deny(missing_debug_implementations)]
#![deny(missing_docs)]
#![cfg_attr(test, allow(dead_code))]
#![cfg_attr(test, deny(warnings))]

//! Libdata re-exports public interface from [datacore],
//! defines async [CoreIterator],
//! defines interface for managing collection of [Cores],
//! and specifies [replication] over [protocol].

pub use datacore::{
    Core, RandomAccess, BlockSignature, Signature,
    MAX_CORE_LENGTH,
};

mod key;
pub use key::{
    Keypair, PublicKey, SecretKey, DiscoveryKey,
    generate_keypair, derive_keypair, discovery_key,
};

mod iter;
pub use iter::CoreIterator;

mod cores;
pub use cores::Cores;

pub mod replication;
