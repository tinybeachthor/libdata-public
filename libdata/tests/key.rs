use quickcheck::{quickcheck, TestResult};
use insta;

use libdata::{generate_keypair, derive_keypair, SecretKey};

#[test]
fn key_can_generate() {
    generate_keypair();
}

#[test]
fn key_can_derive() {
    let keypair = generate_keypair();
    derive_keypair(&keypair.secret, "hello");
}

quickcheck! {
    fn key_same_key_different_names(a: String, b: String) -> TestResult {
        if a == b {
            return TestResult::discard()
        }

        let main = generate_keypair();
        let a = derive_keypair(&main.secret, &a);
        let b = derive_keypair(&main.secret, &b);

        TestResult::from_bool(a.to_bytes() != b.to_bytes())
    }

    fn key_different_key_same_name(name: String) -> bool {
        let a = generate_keypair();
        let b = generate_keypair();
        let a = derive_keypair(&a.secret, &name);
        let b = derive_keypair(&b.secret, &name);

        a.to_bytes() != b.to_bytes()
    }

    fn key_same_key_same_name(name: String) -> bool {
        let main = generate_keypair();
        let a = derive_keypair(&main.secret, &name);
        let b = derive_keypair(&main.secret, &name);

        a.to_bytes() == b.to_bytes()
    }
}

const SECRET_KEY_BYTES: [u8; 32] = [
    157, 097, 177, 157, 239, 253, 090, 096,
    186, 132, 074, 244, 146, 236, 044, 196,
    068, 073, 197, 105, 123, 050, 105, 025,
    112, 059, 172, 003, 028, 174, 127, 096,
];

#[test]
fn key_secret_key_bytes_have_not_changed() {
    insta::assert_debug_snapshot!(SECRET_KEY_BYTES);
}

#[test]
fn key_snapshot_1() {
    let main = SecretKey::from_bytes(&SECRET_KEY_BYTES).unwrap();
    let keypair = derive_keypair(&main, "hello");
    insta::assert_debug_snapshot!(keypair.to_bytes());
}

#[test]
fn key_snapshot_2() {
    let main = SecretKey::from_bytes(&SECRET_KEY_BYTES).unwrap();
    let keypair = derive_keypair(&main, "hello2");
    insta::assert_debug_snapshot!(keypair.to_bytes());
}

#[test]
fn key_snapshot_3() {
    let main = SecretKey::from_bytes(&SECRET_KEY_BYTES).unwrap();
    let keypair = derive_keypair(
        &main,
        "a very long string as a key name should not break the key derive, \
        it should just work without any issues, this is just testing it, \
        to be sure"
    );
    insta::assert_debug_snapshot!(keypair.to_bytes());
}
