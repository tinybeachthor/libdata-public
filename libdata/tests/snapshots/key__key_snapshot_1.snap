---
source: libdata/tests/key.rs
expression: keypair.to_bytes()

---
[
    93,
    98,
    222,
    176,
    190,
    195,
    120,
    66,
    48,
    234,
    225,
    90,
    182,
    230,
    127,
    218,
    48,
    127,
    92,
    187,
    238,
    141,
    123,
    31,
    177,
    55,
    164,
    25,
    222,
    47,
    174,
    183,
    97,
    107,
    208,
    196,
    61,
    246,
    133,
    252,
    52,
    213,
    6,
    217,
    59,
    204,
    226,
    111,
    126,
    68,
    212,
    103,
    63,
    7,
    24,
    88,
    18,
    19,
    151,
    112,
    74,
    228,
    103,
    159,
]
